Merge XGS1210 support and more RTL9300 features

bkobl opened the pull request:
This PR improves the feature support for the RTL9300 platform, specifically it adds:
- SFP+ support for fibre and ethernet modules (10G and 1G)
- Improved RTL8226 and RTL8218d support for the XGS models
- LED setup for the Zyxel XGS models without the need for u-boot
- VLAN fixes
- EEE improvements on all platforms and all supported PHYs
- EEE support for the 1G and 2.5G ports of the XGS 1210-12 and for the 1G ports on the XGS1250
- LDPS support for all platforms

These features are then used to add support for the Zyxel XGS1210-12, which has complete HW support (1GBit, 2.5GBit ports, SFP+, EEE, LDPS)

Commit:
realtek: fix CONFIG_MDIO_SMBUS dependency on RTL83xx

CONFIG_SFP depends on CONFIG_MDIO_SMBUS now and this depends on
CONFIG_I2C_SMBUS.
This created the following warnings on the RTL83xx platforms:
WARNING: unmet direct dependencies detected for MDIO_SMBUS
  Depends on [n]: NETDEVICES [=y] && MDIO_DEVICE [=y] && MDIO_BUS [=y] && I2C_SMBUS [=n]
  Selected by [y]:
  - SFP [=y] && NETDEVICES [=y] && PHYLIB [=y] && I2C [=y] && PHYLINK [=y] && (HWMON [=y] || HWMON [=y]=n)

We remove the overall dependency to CONFIG_MDIO_SMBUS from CONFIG_SFP and instead add
it to CONFIG_I2C_RTL9300 since only the RTL93xx platforms make use of the SMBus capable
I2C controller.

Reported-by: Hauke Mehrtens <hauke@hauke-m.de>
Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Introduce SoC specific Switch IRQ settings

Add SoC specific switch IRQ settings for media change and monitoring
of the SerDes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add support for Zyxel XGS1210-12

The Zyxel XGS1210-12 is an 8x 1 GBit, 2x 2.5 GBit NBaseT ports and 2 SFP+ ports switch

Hardware:
	- RTL9302 SoC: MIPS 34Kc @ 800MHz
	- 128 MByte DRAM
	- 16MBit NOR Flash MX25L12833F
	- RTL8218D 8x GBit PHY
	- 2x RTL8226 2.5 GBit PHYs
	- 2x SFP+ Slots
	- 1 Reset button
	- Externally accessible UART header at the right side:
		Top-down: Vcc (3.3V), TX, RX, GND 115200 8N1
	- System LED

Installation:
	Connect to the Serial Port, power on the device
	Interrupt boot by pressing <ESC>
	Enable network by entering
	 > rtk network on
	Upload initramfs via tftp into RAM:
	 > tftpboot 0x84f00000 192.168.1.100:openwrt-realtek-rtl930x-zyxel_xgs1210-12-initramfs-kernel.bin
	Start initramfs
	 > bootm
	Once OpenWRT is running, install the squashfs image to flash.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Use correct spi compatibility string on RTL93xx

The SPI implementation of the RTL93xx SoCs differs. Use the correct
compatiblity string in order to allow the driver to do the right thing.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: add chip select and parallel io support to SPI driver

This adds support to do a chip select and use parallel IO to the
Realtek SPI driver. RTL83xx supports up to 4 different chip select
lines, whereas RTL93xx supports only 2 different CS lines.
The RTL93xx is supported with up to 4 bit parallel IO, RTL83xx
only supports dual IO.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Only include used ports in IRQ masks

Use only used ports in interrupt masks for SDS/link faults
as otherwise these IRQs constantly fire.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: improve error handling for illegal SMI bus

Improves the error handling in case of an illegal SMI bus
number.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Adapt MAC-PHY connection for HiSGMII ports

On HiSGMII ports of the RTL9300 (used for the RTL8226), the
link between MAC and PHY is not statically HISGMII. Instead
for speeds lower than 2.5G, SGMII is used. Detect this
change, reconfigure the SerDes to use that mode and then
recalibrate that link.

Also remove unused code from rtl83xx_phylink_mac_link_state
which handled a case for the RTL9300 speed settings.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Add support for SFP+ ports on RTL9300

This adds support for SFP and SFP+ modules in RTL9300
based switches. For this we add configuration code
for the RTL9300 SerDes in order to support 1000Base-X
and 10GR phy modes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: handle LED setup for 2.5 GBit ports

This adds handling of the LEDs on 2.5GBit ports as found
on the Zyxel XGS1210-10 using RTL8226 PHYs.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: cleanup of RTL930x switch ISR

This cleans up the ISR of RTL930x switches. It also adds limits
to link failure IRQs which are useful for debug purposes
but can flood the system.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: fix DSA handling of SFP+ ports on RTL9300

This fixes handling of the SFP+ ports in DSA by correcting
the output of the link-state routine.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: fix handling of USXGMII links on RTL9300

USXGMII links are 10G links and not also 2.5GBit links.
While at it we disable RX symbol error IRQs and PHY-update
IRQs as we are not able to do something sensible with them.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix SFP+ port in Zyxel XGS1250-12

The SFP+ port is a managed port which does not have a fixed 10Gbit link
but can also hold 1000Base-X modules. Fix this in the .dts.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Use algorithm to calculate LED counts, instead of heuristics

This counts the number of leds in a led set and sets the led-count
for a given port based on its configured led-set instead of the
type of port.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix stripping of VLAN tags on RTL93xx

On the RTL93xx platforms, the VLAN tags were stripped from egressing
and ingressing packets on a port. Keep all these tags intact.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix mac port up/down configuration on RTL9300

The mac link configuration on the RTL9300 was based on a misunderstanding
of how the registers worked. Fix this.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Add port mirroring support for RTL93xx

This adds support for port mirroring on the RTL93xx platforms.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: reduce chattiness of the SDS configuration

Now that calibration of the fiber modes is understood, remove some of the
chattiness.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: fix reset register access

The reset register on RTL93xx not mereley have bits to execute
a reset of a hardware component, but also configuration bits for
reset procedures. Keep them during executing a reset.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix bugs in MMD-register access on RTL838x

This fixes 2 bugs in MMD access on the RTL838x:
	- We need to preserve the state of the partk-page because the
	  normal PHY r/w functions expect a correct setting
	  in RTL838X_SMI_ACCESS_PHY_CTRL_1
	- b53202a8c3f728c348c5376e5b5fb36af7c37744 introduced
	  calling the hw MMD access functions with mdiobus_c45_devad(regnum)
	  so now we need to invert this op via mdiobus_c45_regad(reg)

This was tested with MMD access register R/W for EEE functionality
on the Zyxel GS1900-10HP.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: improve EEE handling

Instead of setting eee->advertised at the MAC-layer we leave this now
to the PHYs to set correctly. We also introduce reading out the LPI
(Low Power Idle) timer.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: fix hw route lookup

Fix various bugs found in the hw route lookup function.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Add support for RTL8214QF

This adds support for the Realtek RTL8214QF PHY, which has a single 5Gbit
QSGMII link and provides 4 1000BX fiber connections for use with SFP
ports. The RTL8214QF PHY is also the HW base for the 10GBit SFP+
PHYs found in the RTL8396 and RTL93xx SoCs.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Improve RTL9300 SerDes initialization

This improves RTL9300 SerDes initialization, including configuration
settings for different link modes to the MAC.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Add configuration support for the RTL8226

This adds support for the configuration of the RTL8226 PHYs, including
support for HiSGMII and SGMII connections used with the RTL9300 SerDes
for 2.5GBit and 1GBit modes. Setup includes TX configuration and RX
calibration of the PHY-MAC link. We provide for MDIO pin swapping as found in the
Zyxel XGS1210 switch.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Improve RTL8218d configuration

This improves the RTL8218d configuration to make use of the
new setup of the Realtek RTL9300 SerDes including calibration support.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: remove unused PHY registers

These PHY register names are no longer used.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Improve SMI polling configuration based on .dts

Improves error detection of the association of smi addresses to ports
on RTL93xx. Allows empty smi addresses for ports using a PHY linked to an
internal SerDes.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Improve EEE suport in PHY drivers

Improve the EEE support in the PHY drivers of the RTL8218B/D/8826:
 - calculate eee->advertised from MMD values
 - enable/disable EEE via MDIO_MMD_AN.MDIO_AN_EEE_ADV where possible
 - support for 2.5 GBit EEE advertisment/LP report on RTL8226

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+	compatible = "zyxel,xgs1210-12", "realtek,rtl838x-soc";
+	model = "Zyxel XGS1210-12 Switch";
+
+	// mw.l 0xb8003308 0x00200000 active low toggles PHY reset
```
I suppose that wasn't to be there?

```suggestion
```

bkobl replied:
No, especially because then people want you to put a PHY-Reset property somewhere...
This was more to keep track of GPIOs. Currently I know:
`
Board GPIO
============================
Device  Pin  Direction  Default  Current
------- ---- ---------- -------- --------
EXT     8    OUT (OUT ) 0        1       LED: Power
EXT     10   OUT (OUT ) 1        1       LED: Mode
INT     22   IN         1        1       BUTTON: Led-Mode
INT     9    IN         1        1       BUTTON: Reset
INT     12   IN         1        1       BUTTON: Restore
INT     23   OUT        1        1       
EXT     23   OUT (OUT ) 1        1       RESET
EXT     24   OUT (OUT ) 1        1       
EXT     25   OUT (OUT ) 1        1       MAX3421: RESET, active low
INT     15   OUT        0        0       
INT     16   OUT        0        0       
INT     17   IN         0        0       
INT     18   OUT        0        0       
INT     19   IN         0        1      MAX3421: IRQ 
INT     8    OUT        0        0       
INT     13   OUT        0        1      POE: SCL 
INT     14   OUT        0        1      POE: SDA
EXT     0    IN  (IN  ) 1        0      SFP: LOS P24       
EXT     1    IN  (IN  ) 1        1      SFP: LOS P25 
EXT     2    IN  (IN  ) 1        0      SFP: LOS P26
EXT     3    IN  (IN  ) 1        1      SFP: LOS P27
EXT     4    IN  (IN  ) 1        0      SFP: MOD-DEF0 P24
EXT     5    IN  (IN  ) 1        1      SFP: MOD-DEF0 P25
EXT     6    IN  (IN  ) 1        0      SFP: MOD-DEF0 P26
EXT     7    IN  (IN  ) 1        1      SFP: MOD-DEF0 P27
EXT     13   IN  (OUT ) 1        1      SFP: SDA P24
EXT     14   IN  (OUT ) 1        1      SFP: SDA P25       
EXT     21   IN  (OUT ) 1        1      SFP: SDA P26 
EXT     22   IN  (OUT ) 1        1      SFP: SDA P27 
EXT     29   OUT (OUT ) 1        1      SFP: SCL P24, P25       
EXT     30   OUT (OUT ) 1        1      SFP: SDA P26, P27 
EXT     19   IN  (OUT ) 1        1      FAN: SDA
EXT     20   OUT (OUT ) 1        1      FAN: SCL 
EXT     28   OUT (OUT ) 1        1      FAN: SPEED???
EXT     31   IN  (IN  ) 0        0       
EXT     32   IN  (IN  ) 0        1       
EXT     33   IN  (IN  ) 0        0
`

This will be removed as suggested.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+		#size-cells = <0>;
+		i2c-parent = <&i2c0>;
+
+		/* i2c of the left SFP cage: port 11 */
```
These comments tend to be tricky, the left on the front, or the left as seen from the rear when looking from the SOC/PCB PoV?
```suggestion
 		/* i2c of the left SFP+ cage as seen from the front: port 11 */ 
```

bkobl replied:
Will be updated as suggested.

bkobl replied:
Will be fixed as suggested.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+		};
+	};
+
+	sfp0: sfp-p11 {
```
is there any difference between sfp and sfp+ in terms of compatible/naming?

bkobl replied:
There is no sfp+ for the kernel. It is just 10Gbit capable SFP. And `compatible = "sff,sfp+";`does not exist, for example.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+		tx-disable-gpio = <&gpio0 15 GPIO_ACTIVE_HIGH>;
+	};
+
+	// Reset-button: GPIO 22
```
```suggestion
```

bkobl replied:
Will do.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+
+	led_set: led_set@0 {
+		compatible = "realtek,rtl9300-leds";
+		led_set0 = <0x0000 0xffff 0x0a20 0x0b80>; // LED set 0: 1000Mbps,  10/100Mbps
+		led_set1 = <0x0a80 0x0a20 0x0a08 0x0a08>; // LED set 1: 100MBit, 1000MBit, 2.5G
+		led_set2 = <0x0000 0xffff 0x0a20 0x0a01>; // LED set 2: 1000MBit, 10GBit
```
```suggestion
		led_set0 = <0x0000 0xffff 0x0a20 0x0b80>; /* LED set 0: 10Mbit,  100Mbit, 1Gbit */
		led_set1 = <0x0a80 0x0a20 0x0a08 0x0a08>; /* LED set 1: 100Mbit, 1000Mbit, 2.5Gbit */
		led_set2 = <0x0000 0xffff 0x0a20 0x0a01>; /* LED set 2: 1000Mbit, 10Gbit */
```

bkobl replied:
Will fix.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+			rtl9300,smi-address = <0 7>;
+		};
+
+		phy24: ethernet-phy@24 {
```
```suggestion
		/* External single port 1Gbit/2.5Gbit  RTL8226 PHY */
		phy24: ethernet-phy@24 {
```

bkobl replied:
Will do.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+			sds = < 7 >;
+		};
+
+		/* SFP Ports */
```
```suggestion
		/* SFP+ Ports */
```

bkobl replied:
Will do.

oliv3r commented on target/linux/realtek/image/rtl930x.mk:
```diff
@@ -1,5 +1,18 @@
+  ZYXEL_VERS := ABTY
+  DEVICE_VENDOR := Zyxel
+  DEVICE_MODEL := XGS1210-12
+    IMAGE_SIZE := 13312k
```
indentation seems to be off here; not sure what's going on though; normally these are 1 tab indented; but here it's 2 spaces, except for this one.

bkobl replied:
Oops. Will be fixed

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
These tend to be dual licensed, with also MIT, do we not want this here as well?

bkobl replied:
Will do.

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/dsa.c:
```diff
@@ -1024,11 +1024,14 @@ static int rtl83xx_port_enable(struct dsa_switch *ds, int port,
 	pr_debug("%s: %x %d", __func__, (u32) priv, port);
```
You forgot to sign-off this commit ;)

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
```
```suggestion
		partitions: partitions {
```
Would you mind adding an alias for this? It's not clean; as we should have a shared dtsi; but the XGS1010-12 has a different partition layout, and to support that, I'd prefer to only override the actual partitions, and keep the rest of the configuration here for the time being. If the actual chip ever changes, where we require different frequencies, we can talk about aliassing the flash or even the SPI bus ;)

bkobl replied:
Will do.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
What is this frequency based off on? According to the datasheet, our chip has a max freq of 133MHz, so usually this property indicates this.

bkobl replied:
No, this is limited by the SPI frequency of the SoC. The datasheet actually says it is fixed to 50MHz for the RTL9300 and the RTL838x. On the other hand, the MAX3421 can only run with a maximum frequency of the SPI bus of 26MHz, and that is what I see e.g. on the Netgear GS728TPv2, where I was able to measure. So I believe the actual clock frequency is probably only 25MHz. The good news is that the SPI driver does not even try to set a different frequency divider than u-boot (which in the SDK is a relatively high number, definitely not 1) and there is no way to change it via the driver API. So this property is entirely bogus. But I will change it to 50MHz as suggested by the datasheet as the maximum SPI frequency for both SoC and Flash.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+&spi0 {
+	status = "okay";
+	flash@0 {
+		compatible = "jedec,spi-nor";
```
```suggestion:
		compatible = "macronix,mx25l12833f", "mxicy,mx25l12833f", "jedec,spi-nor";
```

bkobl replied:
Did you test that? It might do fancy stuff like parallel IO. The updated SPI driver would support that, now, and I did not test :-(

oliv3r replied:
Tested, works fine (it's unused atm in the code anyway) ...

bkobl commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1210-12.dts:
```diff
@@ -0,0 +1,326 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		/* External RTL8218D PHY */
```
Will fix.

oliv3r commented:
So not sure what you mean, the commented U-Boot memory dump command 
speaks of how to reset a PHY, so not sure how this relates to your 
explanation of found GPIO's :)

But we should certainly put those in the wiki; these are the GPIO's for 
the XGS1210-12 yeah? Do you have this device as well? The wiki doesn't 
say you do :p

oliv3r commented:
Nope, I did not test it yet ;) but I only now am able to do my own 
builds and test those; so I will certainly do it.

The way I know this works, the flash chip driver (not the SPI-NOR) will 
look for this compatible, it's likely not to exist, and that's fine, and 
will fall back to the jdec driver, which will auto-discover most 
parameters. But, the devicetree is to describe hardware, and this is in 
the end the chip that we use, so if there's some quirck in the actual 
flash chip you now have a compatible to fix it. Years ago, the idea was 
you define the DTS once, and never change it. The premisis made sense, 
you put the bootloader and DTB in the eeprom; where you don't really 
change it. However, we have learned since, that device's aren't perfect 
because everything evolves.

oliv3r commented:
I get that, but we are in control of the name, so we could have considered:

sfp0: sfp+-p11

for example, I agree though that that doesn't make it much better (the + 
character is a bit annoying). But I'm also not familiar with sfp namings 
in the kernel currently.

oliv3r commented:
Right, so as I understand it, you set the max-frequency on multiple 
levels, but ignoring everything above the SPI bus level (so ignoring 
cpu, soc, pll etc frequencies that in the end are all of influence), you 
define the max frequency of the SPI bus; the default is set in (our 
case) 930x.dtsi. This indicates the max freq of the bus. Now, if the 
hardware design somehow forces a restriction of the bus, poor signal 
integrity for example, you'd lower this on the spi bus level.

Next, each of the peripherals also have components on them, sometimes 
just one, sometimes many (with chip selects/i2c addresses). There each 
node has its own max-freq setting. So setting it to max freq. of the 
actual chip as per datasheet makes sense.

In the end, the kernel will do the right and pick the appropriate freq. 
for the combination, so if spi bus max is 25MHz, and the chip max bus 
freq is 133MHz, then we'd end up still only speaking at 25MHz. This 
becomes more important if there's several peripherals on the bus with 
different speeds.

bkobl commented:
The mw.l u-boot command actually sets a GPIO output pin to a value that resets the PHY. 0xb8003308 is the address of the register that sets the GPIO pins to a certain output value. mw.l writes 0x00200000 into that register, so pulls one GPIO up. It is actually GPIO 21. But the PHY driver would not know what to do with it at the moment, so I left that out.

Please ignore the GPIO overview. I messed up and that was for the wrong device. The correct table is:
```
============================
 Board GPIO
============================
Device  Pin  Direction  Default  Current  Purpose
------- ---- ---------- -------- -------- --------
INT     0    OUT        0        0        SYSTEM-LED
INT     6    IN         1        1        HW_VER_BIT0
INT     7    IN         1        0        HW_VER_BIT1
INT	8				  SFP0/SFP1: SCK
INT	9				  SFP0: SDA
INT	10				  SFP1: SDA
INT     11   OUT        0        0        SFP0: TX_DIS_FX_0
INT     12   IN         0        0        SFP0: MOD-DEF 
INT     13   IN         0        0        SFP0: LOS
INT     14   IN         0        0        
INT     15   OUT        0        0        TX_DIS_FX_1
INT     16   IN         0        1        SFP1: MOD-DEF
INT     17   IN         0        1        SFP1: LOS
INT     20   IN         0        1
INT     21   OUT        1        1        Reset_PHY
INT     22   IN         1        1        Reset_Button
```
What input pins 14 and 20 do is still unknown.
I have that device (although at the moment I only have a XGS1010-12 around), feel free to add the above to the wiki.

bkobl commented:
> The way I know this works, the flash chip driver (not the SPI-NOR) will look for this compatible, it's likely not to exist, and that's fine, and will fall back to the jdec driver, which will auto-discover most parameters.

I am happy to take any compatible that fits better and works. The driver gets parallel IO support in this PR, but this was not tested on this platform, it was for a different purpose. The flash chip does not seem to be known to Linux 5.10 and seems also not be in the many flash patches in OpenWRT. So the settings are probably very conservative. I don't want to break things at this point, just to use the perfect compatible string. Particularly, I don't have a protocol analyzer at hand at the moment to quickly debug parallel SPI at 50MHz.

bkobl commented:
> sfp0: sfp+-p11

Let's not do this and keep it simple. An SFP port is an SFP port even if it does 10GBit and could also marketed as SFP+.

bkobl commented:
> In the end, the kernel will do the right and pick the appropriate freq. for the combination, so if spi bus max is 25MHz, and the chip max bus freq is 133MHz, then we'd end up still only speaking at 25MHz. This becomes more important if there's several peripherals on the bus with different speeds.

OK. Fixed to 133MHz.

oliv3r commented:
I'll test the compatible, but we might actually need it. The latest master doesn't refer to the string, so should not touch anything.

The reason I say we might nee dis, is that there's an ancient part, the mx25l12805d, that uses the exact same identifier bits, see here.  https://elixir.bootlin.com/linux/v5.19-rc7/source/drivers/mtd/spi-nor/macronix.c#L63

![image](https://user-images.githubusercontent.com/1245341/180601012-9db2351a-2086-43cb-b3e5-c48909a7e7a5.png)

So U-Boot/Linux not adding support for it, is understandable, because they already support our chip, using this old one. However, and that's the tricky bit, the old chip does not support all the fancy features we support. E.g. quad SPI access for example. If you rely on the jedec bits to determine the capabilities of our chip, you get a single SPI chip.

So you cannot detect which chip you use, you cannot detect if you have to do QSPI or not (unless you hardcode it). Having this compatible means that (in the future) we can add the logic needed to support QSPI mode etc by using the macro's such as: `SPI_NOR_QUAD_READ` which we can't right now :(

I have not seen these reuse of ID's before, and how to handle it, so I'd have to look at some examples. But you'd end up with some logic 'if compatible == <ours> then set flags, else use the common denominator (e.g. the slowest chip).

But as you said, lets do this later for sure!

bkobl commented:
> The reason I say we might nee dis, is that there's an ancient part, the mx25l12805d, that uses the exact same identifier bits, see here. https://elixir.bootlin.com/linux/v5.19-rc7/source/drivers/mtd/spi-nor/macronix.c#L63

Indeed. I updated the compatible string and pushed the PR again. Great that you figured this out!

oliv3r committed:
dts: rtl930x: Add shared base for xgs12xx series

The XGS1250-21, XGS1210-12 and XGS1010-12 are very similar, with very
minor differences. Lets have a shared devicetree to make maintaining
multiple devices with a common base easier.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
dts: rtl930x: Migrate XGS1250-12 to common devicetree

With the newly introduced shared devicetree for the XGS12xx series, also
switch the XGS1250-12 to use the shared devicetree.

While here, clean up a small ugliness, without changing actual
functionality.

Note, that due to the weird I2C muxing, there's some overlap there that
will be addressed in the future with a proper I2C driver that properly
handles this.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: add support for Zyxel XGS1010-12

The Zyxel XGS1010-12 is an 8x 1 GBit, 2x 2.5 GBit NBaseT ports and
2 SFP+ ports switch. It is identical to the Zyxel XGS1210-12, except for
the partition layout which is probably caused by the basic firmware
coming from the SDK with the default dual partition layout. To ensure
backwards compatibility, it is best to keep the original partition
layout.

If a larger, single firmware partition, is desired, one can still
install the XGS1210-12 firmware onto the switch.

In addition to the above, the XGS1010-12 lacks a physical reset button,
but all circuitry is in place to support this, including the whole in
the metal housing. As such, we keep the dts entry for the button, to
allow the user to solder on a button, or short the pins, if they so
desire.

As this switch does not feature any form of management or firmware
updateability, one must use the serial UART either through XYZ-modem
(untested) or tftp. The instructions to do so are identical as with the
XGS1250-12.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

svanheule commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl930x.c:
```diff
@@ -2450,6 +2450,8 @@ static void rtl930x_led_init(struct rtl838x_switch_priv *priv)
 		if (priv->ports[i].phy_is_integrated)
 			v = 0x1;
+		if (priv->ports[i].is2G5)
+			v = 0x3;
```
Why on earth would the number of status LEDs on a port depend on the speed of said port? How many LEDs are available for a specific port is entirely up to the vendor to decide and is independent of de capabilities of a specific port.

bkobl replied:
Because higher speeds are indicated by more leds for example???
Yes, this is heuristics. But it works on all known RTL9300 devices and makes a lot of sense.

svanheule replied:
https://www.netgear.com/business/wired/switches/plus/ms108eup/ has RTL930x, and two LEDs per 2.5G port.

oliv3r replied:
As far as I can tell, https://www.zyxel.com/us/en/products_services/12-Port-Web-Managed-Multi-Gigabit-Switch-with-2-Port-2-5G-and-2-Port-10G-SFP--XGS1210-12/ has 2 RGB led's for the two 2.5G ports, so 3 (1x RGB) leds per port. This can be seen on the pcb traces easily as well; 
![image](https://user-images.githubusercontent.com/1245341/181563958-c6c8e34e-761b-4746-b303-b89bc1b595d4.png)
you can see 4 traces, via's and resistors going to the upper and 4 to the lower port led.

I haven't played with the leds in detail yet (will do soon) but from the stock firmware bootup sequence, they show 2 colors per normal led (amber + green) and on the 2.5G they show 'ugly white' and green, to ensure they 'turn on all' leds.

I haven't checked the SFP cages leds, but they blink with two solid colors afaik, blue and something else ...

svanheule replied:
The wiring scheme for the XGS1250-12 is documented on the switch wiki: https://svanheule.net/switches/zyxel_xgs1250-12#port_numbers

It's a weird wiring scheme though. I don't even think you could properly represent all 5 color component in a DT binding.

bkobl replied:
>It's a weird wiring scheme though. I don't even think you could properly represent all 5 color components in a DT description.

They use multiple LEDs on at the same time, like Pink is red plus blue. My 7 year old daughter commented on this, btw and has made a technical request to only have 5GBit connections at home ;-)

I am having the same issue with the GS728TPv2 status LEDs, they are also always double, and e.g. an orange device status because of PoE is created by a green LED plus a red-LED. In the DT I only add the actual LEDs, i.e. red and green.

svanheule commented:
The patch that modifies the SPI driver needs to be submitted upstream for review, or dropped from this series. This is the wrong place to be spending reviewing changes to upstream drivers.

bkobl commented:
There are literally thousands of patches to upstream drivers in OpenWRT, even in the Realtek target, to name just the last ones:
```
target/linux/realtek/patches-5.10$ ls
[...]
711-net-phy-add-an-MDIO-SMBus-library.patch
712-net-phy-sfp-add-support-for-SMBus.patch
713-v5.12-net-dsa-configure-better-brport-flags-when-ports-lea.patch
800-net-mdio-support-hardware-assisted-indirect-access.patch
```
The argument you make cannot be valid.

svanheule commented:
Surely it is a valid argument. AFAIK the SPI driver does not provide any features required by OpenWrt-only code, nor does it provide any such features itself. The patch is also not a hot-fix, which would also warrant direct inclusion in OpenWrt in my opinion. That's why I consider it a reasonable request to ask for the eligible changes to be submitted upstream for review. If there are inconsolable differences in opinions that would make it impossible to move forward upstream, then we can consider to merge these changes only here.

robimarko commented:
I agree with @svanheule changes to upstream drivers, need to go upstream first.
Exceptions would probably be hotfixes where it's more important to get it fixed.

Otherwise, we get to the situation where there are tons of patches to upstream code that were newer sent upstream.
BTW, why weren't MDIO SMBus and SFP SMBus ever sent upstream, as those look like generic improvements for
everybody to enjoy?

bkobl commented:
> Surely it is a valid argument. AFAIK the SPI driver does not provide any features required by OpenWrt-only code, nor does it provide any such features itself. The patch is also not a hot-fix, which would also warrant direct inclusion in OpenWrt in my opinion.

The SPI driver was written by an author unaffiliated with OpenWRT and provides support for devices not supported by OpenWRT. The changes done would possibly affect such devices, none of which I can test on. In particular the CS bits are critical, here. Only a tiny minority of patches in OpenWRT are hot fixes. These SPI fixes specifically add device compatibility strings that are needed on the RTL93xx platforms specific to OpenWRT at this point plus support for these platforms.

>BTW, why weren't MDIO SMBus and SFP SMBus ever sent upstream, as those look like generic improvements for
everybody to enjoy?
These patches, developed together with @dangowrt and largely based on work by people outside of OpenWRT are not complete, but specific to some devices on OpenWRT. In particular, there are issues detecting PHYs on I2C busses in SFP ports that are shared with device on the same bus but not an SFP phy. This would need greater changes to the Linux SFP code to be widely useful, and probably the reason they were not already upstreamed by the original authors.

bkobl commented:
I rebased in order to accomodate for conflicting new includes in the PHY driver and because changes to the hw-based MMD register access wait-loops conflicted with a fix to how this register access is now done, which however no longer required these wait loops at the beginning of the hw setup for the call.

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1010-12.dts:
```diff
@@ -0,0 +1,27 @@
+	#size-cells = <1>;
+
+	partition@300000 {
+		label = "firmware";
+		reg = <0x300000 0x680000>;
+		compatible = "openwrt,uimage", "denx,uimage";
```
```suggestion
```

the only reason we overwrite this partition, is because of the ih-magic. If anybody has any idea how to put the content in some variable, or alias a property; or derivce it some other way from, that would be much nicer :)

oliv3r replied:
```suggestion
```

the only reason we overwrite this partition, is because of the ih-magic. If anybody has any idea how to put the content in some variable, or alias a property; or derivce it some other way from, that would be much nicer :)

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1010-12.dts:
```diff
@@ -0,0 +1,27 @@
+};
+
+&partitions {
+	compatible = "fixed-partitions";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	partition@300000 {
```
made a little booboo here, this all needs to go:
```suggestion
	partition@300000 {
```
while trying to boot the 1210 firmware on my 1010, I noticed this sloppyness (my development leftover that I hadn't cleaned yet). This led me to find a typo bug in the 12xx dtsi; which this code here actually 'fixes' (for the wrong reasons).

Sorry for not testing all 3 variants on my device!

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs12xx.dtsi:
```diff
@@ -0,0 +1,227 @@
+				label = "jffs2";
+				reg = <0x200000 0x100000>;
+			};
+			partition@b300000 {
```
```suggestion
			partition@300000 {
```
typo! sorry

oliv3r commented on target/linux/realtek/dts-5.10/rtl9302_zyxel_xgs1010-12.dts:
```diff
@@ -0,0 +1,27 @@
+	#size-cells = <1>;
+
+	partition@300000 {
+		label = "firmware";
```
```suggestion
```

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -3730,6 +3943,7 @@ static int rtl8214c_phy_probe(struct phy_device *phydev)
 {
 	struct device *dev = &phydev->mdio.dev;
 	int addr = phydev->mdio.addr;
+	int ret = 0;
```
Just be aware, that your other patch 'improve rtl8218d' also relies on this, so these two patches are 'married' because of this, change.

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -4875,6 +4876,7 @@ static struct phy_driver rtl83xx_phy_driver[] = {
 		.features	= PHY_GBIT_FEATURES,
 		.flags		= PHY_HAS_REALTEK_PAGES,
 		.probe		= rtl8218d_phy_probe,
+		.config_init	= rtl9300_configure_8218d,
```
Note, that within this patch, this does not yet exist, so probably comes from one of your other patches?

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c:
```diff
@@ -3165,13 +3294,295 @@ int rtl9300_sds_cmu_band_get(int sds)
+	return 0;
+}
+
+int rtl9300_configure_8218d(struct phy_device *phydev)
```
don't forget to reference this function in line [4572](https://github.com/openwrt/openwrt/pull/10319/commits/aa0640117cd3ec79faf39485a8b2eb6ec45857ed#diff-f2e3b446d6bf87d07398d90e24463324c5e7a1a033610512f668d1046bfdab25R4572) which you accidentally did in improve rtl8218d patch; i think this function is unused in this patch

bkobl commented:
> The patch that modifies the SPI driver needs to be submitted upstream for review, or dropped from this series. This is the wrong place to be spending time reviewing changes to upstream drivers.

Since the upstream maintainer recently suggested to submit the patches, I just submitted them upstream. The externally visible changes of the patches should be limited to the rtl9300 compatibility string, so for now I suggest we just leave this patch in as a placeholder, so that the .dts of the devices can stay clean.

Neustradamus commented:
A new PR has been created:
- https://github.com/openwrt/openwrt/pull/10453
